File: app/src/main/cpp/wn-steam-client/src/steam_directory.cpp
#include "wn_steam/steam_directory.h"

#include <android/log.h>
#include <curl/curl.h>
#include <nlohmann/json.hpp>

#include <cstring>

namespace wn_steam {

namespace {

constexpr const char* kDirectoryEndpoint =
    "https://api.steampowered.com/ISteamDirectory/GetCMListForConnect/v1/";

constexpr const char* kLogTag = "WnSteamDir";

#define WN_LOGE(...) __android_log_print(ANDROID_LOG_ERROR, kLogTag, __VA_ARGS__)
#define WN_LOGI(...) __android_log_print(ANDROID_LOG_INFO,  kLogTag, __VA_ARGS__)

size_t curl_write_cb(char* ptr, size_t size, size_t nmemb, void* userdata) {
    auto* out = static_cast<std::string*>(userdata);
    const size_t n = size * nmemb;
    out->append(ptr, n);
    return n;
}

CmTransport parse_transport(std::string_view s) noexcept {
    // Steam Directory uses "websockets" (plural). Accept singular too.
    if (s == "websockets" || s == "websocket") return CmTransport::WebSocket;
    if (s == "netfilter")                      return CmTransport::Tcp;
    return CmTransport::Unknown;
}

}  // namespace

SteamDirectoryClient::SteamDirectoryClient() {
    // libcurl global init is documented thread-unsafe but idempotent; the
    // Android Gradle plugin's curl Prefab AAR runs it implicitly via the
    // first easy-init in most loaders. Calling it again is harmless after
    // the first time and necessary if no other consumer pulled it in yet.
    curl_global_init(CURL_GLOBAL_DEFAULT);
}

SteamDirectoryClient::~SteamDirectoryClient() {
    // Note: we intentionally do NOT call curl_global_cleanup here. Other
    // wn-steam-client subsystems also use libcurl and there's no reference
    // counting on the global init. The OS reclaims everything at process
    // teardown.
}

SteamDirectoryResult SteamDirectoryClient::fetch(uint32_t cell_id,
                                                 std::chrono::seconds timeout,
                                                 std::string_view user_agent,
                                                 std::string_view ca_bundle_path) {
    SteamDirectoryResult result;
    result.http_status = 0;

    char url[512];
    // cmtype=websockets so the response excludes TCP CMs we won't use.
    // maxcount=20 mirrors what the official client requests.
    std::snprintf(url, sizeof(url),
                  "%s?cellid=%u&cmtype=websockets&maxcount=20&realm=steamglobal",
                  kDirectoryEndpoint, cell_id);

    CURL* curl = curl_easy_init();
    if (!curl) {
        result.error = "curl_easy_init failed";
        WN_LOGE("%s", result.error.c_str());
        return result;
    }

    std::string body;
    body.reserve(8192);

    std::string ua(user_agent);

    curl_easy_setopt(curl, CURLOPT_URL,            url);
    curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION,  curl_write_cb);
    curl_easy_setopt(curl, CURLOPT_WRITEDATA,      &body);
    curl_easy_setopt(curl, CURLOPT_USERAGENT,      ua.c_str());
    curl_easy_setopt(curl, CURLOPT_TIMEOUT,        static_cast<long>(timeout.count()));
    curl_easy_setopt(curl, CURLOPT_CONNECTTIMEOUT, static_cast<long>(timeout.count()));
    curl_easy_setopt(curl, CURLOPT_FOLLOWLOCATION, 1L);
    curl_easy_setopt(curl, CURLOPT_NOPROGRESS,     1L);

    // TLS — Prefab curl is built against OpenSSL. The default verify
    // settings are on; Android system CAs aren't shipped in the curl AAR
    // sysroot. CURLOPT_CAPATH=/system/etc/security/cacerts (hashed-dir
    // mode) was empirically broken on Android 16 / OnePlus testing, so
    // we instead use CURLOPT_CAINFO pointing at the single-file PEM
    // bundle CaBundleExtractor produced from those same hashed dirs.
    std::string ca_path(ca_bundle_path);
    if (!ca_path.empty()) {
        curl_easy_setopt(curl, CURLOPT_CAINFO, ca_path.c_str());
    }
    curl_easy_setopt(curl, CURLOPT_SSL_VERIFYPEER, 1L);
    curl_easy_setopt(curl, CURLOPT_SSL_VERIFYHOST, 2L);

    const CURLcode rc = curl_easy_perform(curl);

    long http_status = 0;
    curl_easy_getinfo(curl, CURLINFO_RESPONSE_CODE, &http_status);
    result.http_status = static_cast<int>(http_status);

    curl_easy_cleanup(curl);

    if (rc != CURLE_OK) {
        result.error = "curl_easy_perform: ";
        result.error += curl_easy_strerror(rc);
        WN_LOGE("%s (http_status=%ld)", result.error.c_str(), http_status);
        return result;
    }

    if (http_status != 200) {
        result.error = "non-200 HTTP status";
        WN_LOGE("Steam Directory returned HTTP %ld", http_status);
        return result;
    }

    // Parse JSON body. Expected shape:
    //   { "response": { "success": bool, "serverlist": [ { ... }, ... ] } }
    try {
        auto root = nlohmann::json::parse(body);
        const auto& response = root.at("response");

        // `success` in the directory response is an int (1 = OK) per the
        // current schema; accept either int or bool for robustness.
        bool ok = false;
        if (response.contains("success")) {
            const auto& s = response["success"];
            if (s.is_boolean())       ok = s.get<bool>();
            else if (s.is_number())   ok = (s.get<int>() == 1);
        }
        if (!ok) {
            result.error = "directory response: success=false";
            WN_LOGE("%s", result.error.c_str());
            return result;
        }

        if (!response.contains("serverlist") || !response["serverlist"].is_array()) {
            result.error = "directory response: missing serverlist";
            WN_LOGE("%s", result.error.c_str());
            return result;
        }

        const auto& list = response["serverlist"];
        result.servers.reserve(list.size());

        for (const auto& entry : list) {
            CmServer s;

            if (entry.contains("endpoint") && entry["endpoint"].is_string()) {
                s.endpoint = entry["endpoint"].get<std::string>();
            } else {
                continue;  // unparseable, skip
            }
            if (!parse_endpoint(s.endpoint, s.host, s.port)) {
                continue;
            }

            if (entry.contains("type") && entry["type"].is_string()) {
                s.transport = parse_transport(entry["type"].get<std::string>());
            }
            // Phase 1 only handles WebSocket transport.
            if (s.transport != CmTransport::WebSocket) continue;

            if (entry.contains("realm") && entry["realm"].is_string())
                s.realm = entry["realm"].get<std::string>();
            if (entry.contains("dc") && entry["dc"].is_string())
                s.datacenter = entry["dc"].get<std::string>();
            if (entry.contains("load") && entry["load"].is_number_integer())
                s.load = entry["load"].get<int32_t>();
            if (entry.contains("wtd_load") && entry["wtd_load"].is_number())
                s.weighted_load = entry["wtd_load"].get<float>();

            result.servers.push_back(std::move(s));
        }

        WN_LOGI("Steam Directory returned %zu WebSocket servers (cellid=%u)",
                result.servers.size(), cell_id);
    } catch (const nlohmann::json::exception& e) {
        result.error = "json parse error: ";
        result.error += e.what();
        WN_LOGE("%s", result.error.c_str());
        result.servers.clear();
    }

    return result;
}

}  // namespace wn_steam

File: app/src/main/cpp/wn-steam-client/rust/src/steam_directory.rs
use crate::cm_server::{parse_endpoint, CmServer, CmTransport};
use serde_json::Value;
use std::fs;
use std::time::Duration;

pub const DEFAULT_USER_AGENT: &str = "Valve/Steam HTTP Client 1.0";
pub const DIRECTORY_ENDPOINT: &str =
    "https://api.steampowered.com/ISteamDirectory/GetCMListForConnect/v1/";
pub const DEFAULT_TIMEOUT: Duration = Duration::from_secs(10);

#[derive(Clone, Debug, Default, PartialEq)]
pub struct SteamDirectoryResult {
    pub servers: Vec<CmServer>,
    pub error: String,
    pub http_status: i32,
}

impl SteamDirectoryResult {
    pub fn ok(&self) -> bool {
        self.error.is_empty() && !self.servers.is_empty()
    }
}

#[derive(Clone, Debug, Default, Eq, PartialEq)]
pub struct SteamDirectoryClient {
    ca_bundle_path: String,
}

impl SteamDirectoryClient {
    pub fn new(ca_bundle_path: impl Into<String>) -> Self {
        Self {
            ca_bundle_path: ca_bundle_path.into(),
        }
    }

    pub fn ca_bundle_path(&self) -> &str {
        &self.ca_bundle_path
    }

    pub fn build_url(cell_id: u32) -> String {
        format!(
            "{DIRECTORY_ENDPOINT}?cellid={cell_id}&cmtype=websockets&maxcount=20&realm=steamglobal"
        )
    }

    pub fn fetch(&self, cell_id: u32, timeout: Duration) -> SteamDirectoryResult {
        self.fetch_with_user_agent(cell_id, timeout, DEFAULT_USER_AGENT)
    }

    pub fn fetch_with_user_agent(
        &self,
        cell_id: u32,
        timeout: Duration,
        user_agent: &str,
    ) -> SteamDirectoryResult {
        let url = SteamDirectoryClient::build_url(cell_id);
        match self.http_get_text(&url, timeout, user_agent) {
            Ok(response) => {
                SteamDirectoryClient::validate_response(response.http_status, &response.body)
            }
            Err(error) => SteamDirectoryResult {
                error,
                ..Default::default()
            },
        }
    }

    pub fn validate_response(http_status: i32, body: &str) -> SteamDirectoryResult {
        if http_status != 200 {
            return SteamDirectoryResult {
                error: "non-200 HTTP status".to_string(),
                http_status,
                ..Default::default()
            };
        }
        let mut result = parse_directory_response(body);
        result.http_status = http_status;
        result
    }

    pub fn default_timeout() -> Duration {
        DEFAULT_TIMEOUT
    }

    fn http_get_text(
        &self,
        url: &str,
        timeout: Duration,
        user_agent: &str,
    ) -> Result<HttpTextResponse, String> {
        let client = self.http_client(timeout, user_agent)?;
        let response = client
            .get(url)
            .send()
            .map_err(|err| format!("http get: {err}"))?;
        let http_status = response.status().as_u16() as i32;
        let body = response.text().map_err(|err| format!("http body: {err}"))?;
        Ok(HttpTextResponse { http_status, body })
    }

    fn http_client(
        &self,
        timeout: Duration,
        user_agent: &str,
    ) -> Result<reqwest::blocking::Client, String> {
        let mut builder = reqwest::blocking::Client::builder()
            .user_agent(user_agent)
            .timeout(timeout)
            .connect_timeout(timeout);
        if !self.ca_bundle_path.is_empty() {
            let pem =
                fs::read(&self.ca_bundle_path).map_err(|err| format!("read CA bundle: {err}"))?;
            let certs = reqwest::Certificate::from_pem_bundle(&pem)
                .map_err(|err| format!("parse CA bundle: {err}"))?;
            for cert in certs {
                builder = builder.add_root_certificate(cert);
            }
        }
        builder.build().map_err(|err| format!("http client: {err}"))
    }
}

struct HttpTextResponse {
    http_status: i32,
    body: String,
}

pub fn parse_directory_response(body: &str) -> SteamDirectoryResult {
    let mut result = SteamDirectoryResult::default();
    let root: Value = match serde_json::from_str(body) {
        Ok(v) => v,
        Err(e) => {
            result.error = format!("json parse error: {e}");
            return result;
        }
    };
    let Some(response) = root.get("response") else {
        result.error = "directory response: missing response".to_string();
        return result;
    };
    let ok = response
        .get("success")
        .map(|s| {
            s.as_bool()
                .unwrap_or_else(|| s.as_i64().is_some_and(|n| n == 1))
        })
        .unwrap_or(false);
    if !ok {
        result.error = "directory response: success=false".to_string();
        return result;
    }
    let Some(list) = response.get("serverlist").and_then(|v| v.as_array()) else {
        result.error = "directory response: missing serverlist".to_string();
        return result;
    };

    for entry in list {
        let Some(endpoint) = entry.get("endpoint").and_then(|v| v.as_str()) else {
            continue;
        };
        let Some((host, port)) = parse_endpoint(endpoint) else {
            continue;
        };
        let transport = entry
            .get("type")
            .and_then(|v| v.as_str())
            .map(parse_transport)
            .unwrap_or(CmTransport::Unknown);
        if transport != CmTransport::WebSocket {
            continue;
        }
        result.servers.push(CmServer {
            endpoint: endpoint.to_string(),
            host,
            port,
            transport,
            realm: entry
                .get("realm")
                .and_then(|v| v.as_str())
                .unwrap_or("")
                .to_string(),
            datacenter: entry
                .get("dc")
                .and_then(|v| v.as_str())
                .unwrap_or("")
                .to_string(),
            load: entry.get("load").and_then(|v| v.as_i64()).unwrap_or(0) as i32,
            weighted_load: entry
                .get("wtd_load")
                .and_then(|v| v.as_f64())
                .unwrap_or(0.0) as f32,
        });
    }
    result
}

pub fn parse_transport(s: &str) -> CmTransport {
    match s {
        "websockets" | "websocket" => CmTransport::WebSocket,
        "netfilter" => CmTransport::Tcp,
        _ => CmTransport::Unknown,
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn parses_websocket_entries_and_filters_tcp() {
        let json = r#"{
          "response": {
            "success": 1,
            "serverlist": [
              {"endpoint":"ext1-sea1.steamserver.net:443","type":"websockets","realm":"steamglobal","dc":"sea1","load":1,"wtd_load":2.5},
              {"endpoint":"tcp.example.com:27017","type":"netfilter"},
              {"endpoint":"bad","type":"websockets"}
            ]
          }
        }"#;
        let result = parse_directory_response(json);
        assert!(result.error.is_empty());
        assert_eq!(result.servers.len(), 1);
        assert_eq!(result.servers[0].host, "ext1-sea1.steamserver.net");
        assert_eq!(result.servers[0].weighted_load, 2.5);
    }

    #[test]
    fn accepts_boolean_success() {
        let json = r#"{"response":{"success":true,"serverlist":[]}}"#;
        let result = parse_directory_response(json);
        assert!(result.error.is_empty());
        assert!(result.servers.is_empty());
    }

    #[test]
    fn builds_directory_url_and_tracks_ca_bundle() {
        let client = SteamDirectoryClient::new("/cacert.pem");
        assert_eq!(client.ca_bundle_path(), "/cacert.pem");
        assert_eq!(
            SteamDirectoryClient::default_timeout(),
            Duration::from_secs(10)
        );
        assert_eq!(
            SteamDirectoryClient::build_url(123),
            "https://api.steampowered.com/ISteamDirectory/GetCMListForConnect/v1/?cellid=123&cmtype=websockets&maxcount=20&realm=steamglobal"
        );
    }

    #[test]
    fn validates_http_status_before_json() {
        let result = SteamDirectoryClient::validate_response(503, "{}");
        assert_eq!(result.http_status, 503);
        assert_eq!(result.error, "non-200 HTTP status");

        let ok = SteamDirectoryClient::validate_response(
            200,
            r#"{"response":{"success":1,"serverlist":[{"endpoint":"ext1-sea1.steamserver.net:443","type":"websockets"}]}}"#,
        );
        assert_eq!(ok.http_status, 200);
        assert!(ok.ok());
    }
}
